### DISCUSSION TITLE
Echo events continue to be triggered after page change and gives component not found error

### DISCUSSION THREAD

**Discussion by loxi5** [livewire/livewire#7397] [7. Bugs (archived)]
### Livewire version

v3.0.0

### Which PHP version are you using?

PHP 8.2

### Steps To Reproduce

1. Create a full page component including another component which listens to "Laravel Echo" events. This component must also include sub components. Give every component a unique key. In my case, the full page component is messages page. It has messages component and this component has a loop of message components.
2. Visit this page and send a notification. You can see that the event is triggered and working fine. In my case I run a public method, get the new message and prepend to the messages list. So this message is visible on top of the loop as a component.
3. Visit another page **(which does not include the component listening for the events on previous page)**. This page must be visited with **wire:navigate**.
4. Send a notification again. You will see that it will try to trigger the echo event on the previous component. But the event will not be triggered because the component does not exist. In my case it continues listening to the Echo Event on "messages" component but before triggering, it tries to get sub components (looped message components) but it cannot find. Then gives the error: **Component not found: [id]**

**Error log:**

> livewire.esm.js?t=1701181050763:7370 Uncaught (in promise) Component not found: rHpp8x5PjMNn9XzjaSYW
sendRequestToServer @ livewire.esm.js?t=1701181050763:7370
(anonymous) @ livewire.esm.js?t=1701181050763:7290
(anonymous) @ livewire.esm.js?t=1701181050763:7298
setTimeout (async)
bundleMultipleRequestsTogetherIfTheyHappenWithinFiveMsOfEachOther @ livewire.esm.js?t=1701181050763:7297
triggerSend @ livewire.esm.js?t=1701181050763:7289
(anonymous) @ livewire.esm.js?t=1701181050763:7447
(anonymous) @ livewire.esm.js?t=1701181050763:7404
waitUntilTheCurrentRequestIsFinished @ livewire.esm.js?t=1701181050763:7400
requestCall @ livewire.esm.js?t=1701181050763:7445
(anonymous) @ livewire.esm.js?t=1701181050763:7952
(anonymous) @ livewire.esm.js?t=1701181050763:7891
(anonymous) @ livewire.esm.js?t=1701181050763:8105
dispatchEvent @ livewire.esm.js?t=1701181050763:8122
dispatchSelf @ livewire.esm.js?t=1701181050763:8131
(anonymous) @ livewire.esm.js?t=1701181050763:9312
emit @ dispatcher.ts:76
handleEvent @ channel.ts:86
(anonymous) @ pusher.ts:116
emit @ dispatcher.ts:76
message @ connection_manager.ts:262
emit @ dispatcher.ts:76
message @ connection.ts:118
emit @ dispatcher.ts:76
onMessage @ transport_connection.ts:187
socket.onmessage @ transport_connection.ts:205
Show 10 more frames
Show less

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by codevApp** [comment]
Same Problem.

1. First Component with Echo Event Listener, working fine.
2. wire:navigate to another component without the any listener. 
3. Component breaks with: Uncaught Component not found: xxx
= Bug: I think the Listener is still active from the old component.

**Comment by heyjorgedev** [comment]
It happens to me too, I went into a bit on an exploration and I can see that Livewire is only registering the Echo listeners in the [supportsLaravelEcho.js feature](https://github.com/livewire/livewire/blob/2be9b1b9325b56ebe7d22937ea6d5e2a131bfec2/js/features/supportLaravelEcho.js) but it never stops listening to them on navigation.

I think somehow the [whenTheBackOrForwardButtonIsClicked](https://github.com/livewire/livewire/blob/2be9b1b9325b56ebe7d22937ea6d5e2a131bfec2/js/plugins/navigate/index.js#L98C5-L98C40) and the [navigateTo](https://github.com/livewire/livewire/blob/2be9b1b9325b56ebe7d22937ea6d5e2a131bfec2/js/plugins/navigate/index.js#L54C14-L54C24) need to be aware of Laravel Echo and manage the listeners for the according page navigations.

**Comment by SkeyPunyapal** [comment]
facing the same issue

**Comment by MrPunyapal** [comment]
I checked the issue and went through the code...

this will be tricky to handle as if we stop listening to that event that may cause issues with other components that may still want to listen.

I found the solution to stop listening to unwanted echo events manually... 

you can dispatch an event to stop listing to that particular event.

example 

from component

```php
  $this->dispatch('cleanup-echo-event', channelName: 'XYZ', eventName: 'ABC');
```

in js (make sure it is in the parent component or on the layout)
```js
document.addEventListener('cleanup-echo-event', (event) => {

  // If you want to leave the whole channel
  window.Echo.leaveChannel(event.detail.channelName);
  
  // If you want to a particular event on the channel 
  window.Echo.private(event.detail.channelName)
    .stopListening(event.eventName);

});
```

**Reply by calebporzio** [reply]
If someone can submit a failing test for this I'll look into solving

**Comment by calebporzio** [comment]
Ok, I attempted to fix it (PR here: #7833). Can someone pull down `main` in their local app and verify that it fixed it? It's kind of hard to write an automated test for.

(You can up us the main branch by changing the livewire/livewire version in compoer.json to "dev-main" and running "composer udpdate")

Thanks!

**Reply by MrPunyapal** [reply]
Will it work with switching orders.id in components? 

Example I have something like Below and I dynamically changing the order in same component? 🤔

```php
#[On('echo:orders.{order.id},OrderShipped')]
public function notifyNewOrder()
{
    $this->showNewOrderNotification = true;
}
```

**Reply by joshhanley** [reply]
@MrPunyapal nope, otherwise we would need to send the event names on every request which would add overhead.

**Reply by MrPunyapal** [reply]
@joshhanley let's come to normal events...

It's working with normal events right? 🤔

If I switch Id dynamically, it removes the old listener and new one ✔️

So why can't it work with echo events? 🤔

(Just asking to know if I'm missing something)

**Reply by joshhanley** [reply]
@MrPunyapal hmm I don't think it works with normal events either. I may be wrong 🤔 definitely was the case in v2, I didn't think that changed in v3, but I'd have to test it.

**Reply by calebporzio** [reply]
You'll have to test but I don't think that works

**Comment by alexwass-lr** [comment]
I think I may have found an accidental side effect to this change since updating to 3.4.4 @calebporzio.

I have nested lazy components which are listening on a presence channel for [model broadcasted](https://laravel.com/docs/10.x/broadcasting#listening-for-model-broadcasts) events. There is a search on the parent component which will filter these children. (_Think of it like a searchable to-do list (parent), with each to-do item (child) having a listener so it can be refreshed if another user makes a change._)

Hopefully that made sense, it's hard to abstract it out to a simple example.

However, when a child component is removed (by not matching the search query), the [new cleanup](https://github.com/livewire/livewire/pull/7833/files#diff-71be7e35f51d605a8814367f54b7ac1ddf101170c889f2d2befe0d4cd7f60f9eR8788) is called and throws the console error:
```js
Uncaught TypeError: window.Echo[channel_type] is not a function
```

This worked fine before 3.4.4. Now once this error is thrown the whole app seizes and requires a refresh.

I can take a stab at a failing test if you'd like, but like you said I'm not sure how we'd test this...

I would assume based on [the docs](https://laravel.com/docs/10.x/broadcasting#stop-listening-for-events) that `stopListening` can only be used on private channels, so it should instead be:
```js
component.addCleanup(() => {
    window.Echo.leaveChannel(channel)
})
```

If I make that change locally it works as expected - however not sure if that breaks the navigation issue. If it doesn't though, happy to put in a PR if you like?

**Reply by joshhanley** [reply]
@alexwass-lr any chance you could put a breakpoint on that line of code in your browser and trigger it, and report back what the contents of `channel_type` is at that stage, and also what `window.Echo` looks like then too?

Screenshots of the step debugger showing the values should be fine.

**Reply by alexwass-lr** [reply]
Absolutely @joshhanley, hope these help - I tried to expand what I felt relevant but let me know if I've cut off something you needed

<img width="264" alt="Screenshot 2024-02-02 at 10 16 11" src="https://github.com/livewire/livewire/assets/56585741/93648be0-5d4e-4916-ab38-716ae6de6819">
<img width="798" alt="Screenshot 2024-02-02 at 10 16 04" src="https://github.com/livewire/livewire/assets/56585741/d9472d48-f80d-46f7-8a7e-edcb1ae42fe2">

**Reply by joshhanley** [reply]
@alexwass-lr thanks that's great! Yep you are correct about the wrong clean up listener. If you could submit a PR with a fix that'd be great! I'll leave some guidance below.

This line 62 is the one that is broken and yeah should be changed to `window.Echo.leaveChannel...`

https://github.com/livewire/livewire/blob/dc8ad6e08fa207944f26b8821776c156e88b460d/js/features/supportLaravelEcho.js#L62

But this line 47 below is correct, so don't change that. 

https://github.com/livewire/livewire/blob/dc8ad6e08fa207944f26b8821776c156e88b460d/js/features/supportLaravelEcho.js#L47

The reason it is correct, is because you can see it's inside a conditional checking for `'channel', 'private', 'encryptedPrivate'` channel types, all of which exist as methods on the base `Echo` object.

Hope this helps!

**Reply by alexwass-lr** [reply]
Great thanks @joshhanley, I'll pop that in now

**Reply by joshhanley** [reply]
@alexwass-lr thanks! 🙏
